File: shifai-ios/ShifAI/Data/AnalyticsTracker.swift
import Foundation

/// Analytics Tracker — privacy-safe event tracking via Plausible.
/// Zero PII: no user_id, no health data, no device fingerprints.
/// Consent-based: disabled by default, user enables in Settings.
/// Mirrors Android AnalyticsTracker.kt.
final class AnalyticsTracker {

    static let shared = AnalyticsTracker()

    private let plausibleURL = URL(string: "https://plausible.io/api/event")!

    var isEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: "analytics_consent") }
        set { UserDefaults.standard.set(newValue, forKey: "analytics_consent") }
    }

    private init() {}

    // MARK: - Core

    func track(_ event: String, props: [String: String] = [:]) async {
        guard isEnabled else { return }

        var body: [String: Any] = [
            "name": event,
            "url": "app://shifai/\(event.replacingOccurrences(of: "_", with: "/"))",
            "domain": AppConfig.plausibleDomain
        ]
        if !props.isEmpty { body["props"] = props }

        guard let data = try? JSONSerialization.data(withJSONObject: body) else { return }

        var request = URLRequest(url: plausibleURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("ShifAI iOS/\(AppConfig.appVersion)", forHTTPHeaderField: "User-Agent")
        request.httpBody = data
        request.timeoutInterval = 5

        _ = try? await URLSession.shared.data(for: request)
        // Analytics failures are silent — never block user flow
    }

    // MARK: - Convenience Events

    func trackAppLaunched() async {
        await track("app_launched", props: ["platform": "ios"])
    }

    func trackOnboardingCompleted(cycleLengthBucket: String) async {
        await track("onboarding_completed", props: ["cycle_length_bucket": cycleLengthBucket])
    }

    func trackTrackingSaved(symptomCountBucket: String) async {
        await track("tracking_saved", props: ["symptom_count_bucket": symptomCountBucket])
    }

    func trackExportGenerated(template: String, dateRange: String) async {
        await track("export_generated", props: ["template": template, "date_range": dateRange])
    }

    func trackSyncCompleted(conflictBucket: String) async {
        await track("sync_completed", props: ["conflict_count_bucket": conflictBucket])
    }

    func trackError(errorCode: String) async {
        await track("error_occurred", props: ["error_code": errorCode])
    }
}

File: shifai-android/app/src/main/java/com/shifai/data/analytics/AnalyticsTracker.kt
package com.shifai.data.analytics

import android.content.SharedPreferences
import com.shifai.config.AppConfig
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject

/**
 * Analytics Tracker — privacy-safe event tracking via Plausible.
 * Zero PII: no user_id, no health data, no device fingerprints.
 * Consent-based: disabled by default, user enables in Settings.
 */
class AnalyticsTracker(private val prefs: SharedPreferences) {

    companion object {
        private const val KEY_CONSENT = "analytics_consent"
        private const val PLAUSIBLE_URL = "https://plausible.io/api/event"
    }

    var isEnabled: Boolean
        get() = prefs.getBoolean(KEY_CONSENT, false)
        set(value) { prefs.edit().putBoolean(KEY_CONSENT, value).apply() }

    /**
     * Track a privacy-safe event.
     * Only sends if user has given consent.
     */
    suspend fun track(event: String, props: Map<String, String> = emptyMap()) {
        if (!isEnabled) return

        withContext(Dispatchers.IO) {
            try {
                val body = JSONObject().apply {
                    put("name", event)
                    put("url", "app://shifai/${event.replace("_", "/")}")
                    put("domain", AppConfig.plausibleDomain)
                    if (props.isNotEmpty()) {
                        put("props", JSONObject(props))
                    }
                }

                val conn = URL(PLAUSIBLE_URL).openConnection() as HttpURLConnection
                conn.requestMethod = "POST"
                conn.setRequestProperty("Content-Type", "application/json")
                conn.setRequestProperty("User-Agent", "ShifAI Android/${AppConfig.appVersion}")
                conn.doOutput = true
                conn.connectTimeout = 5000
                conn.readTimeout = 5000

                conn.outputStream.use { it.write(body.toString().toByteArray()) }
                conn.responseCode // trigger send
                conn.disconnect()
            } catch (_: Exception) {
                // Analytics failures are silent — never block user flow
            }
        }
    }

    // ─── Convenience Events ───

    suspend fun trackAppLaunched() = track("app_launched", mapOf("platform" to "android"))

    suspend fun trackOnboardingCompleted(cycleLengthBucket: String) =
        track("onboarding_completed", mapOf("cycle_length_bucket" to cycleLengthBucket))

    suspend fun trackTrackingSaved(symptomCountBucket: String) =
        track("tracking_saved", mapOf("symptom_count_bucket" to symptomCountBucket))

    suspend fun trackExportGenerated(template: String, dateRange: String) =
        track("export_generated", mapOf("template" to template, "date_range" to dateRange))

    suspend fun trackSyncCompleted(conflictBucket: String) =
        track("sync_completed", mapOf("conflict_count_bucket" to conflictBucket))

    suspend fun trackError(errorCode: String) =
        track("error_occurred", mapOf("error_code" to errorCode))
}
